File: AnalizadorBigO/src/main/java/com/mycompany/analizadorbigo/BigOAnalyzer.java
package com.mycompany.analizadorbigo;

import java.util.HashMap;
import java.util.Map;

public class BigOAnalyzer {

    private Map<String, Integer> complexitiesCount; // Almacena las complejidades detectadas

    public BigOAnalyzer() {
        this.complexitiesCount = new HashMap<>();
    }

    public String analyzeCode(String code) {
        if (!code.contains("{") || !code.contains("}")) {
            return "El código no parece ser Java. Por favor, ingrese código válido en Java.";
        }

        String[] lines = code.split("\n");
        StringBuilder lineAnalysis = new StringBuilder();
        int loopDepth = 0; // Para detectar la profundidad de bucles anidados

        // Reiniciar conteo de complejidades
        complexitiesCount.clear();

        // Análisis línea por línea
        for (int i = 0; i < lines.length; i++) {
            String line = lines[i].trim();

            if (line.startsWith("for") || line.startsWith("while")) {
                loopDepth++;
                String complexity = getComplexityByDepth(loopDepth);
                lineAnalysis.append("Línea ").append(i + 1).append(": ").append(complexity).append("\n");
                complexitiesCount.put(complexity, complexitiesCount.getOrDefault(complexity, 0) + 1);
            } else if (line.equals("}")) {
                loopDepth = Math.max(0, loopDepth - 1); // Reducir profundidad al cerrar un bloque
            }
        }

        // Resumen total de complejidades
        StringBuilder totalComplexities = new StringBuilder();
        complexitiesCount.forEach((key, value) -> totalComplexities.append(value).append("x ").append(key).append(" + "));
        if (totalComplexities.length() > 0) {
            totalComplexities.setLength(totalComplexities.length() - 3); // Quitar " + " al final
        }

        String simplifiedComplexity = simplifyBigO(complexitiesCount);
        String quality = evaluateCodeQuality(simplifiedComplexity);

        return lineAnalysis.toString()
                + "Complejidad Total: " + totalComplexities + "\n"
                + "Simplificado: " + simplifiedComplexity + "\n"
                + "Calidad del Código: " + quality
                + "\n\nNota: Los bucles anidados incrementan exponencialmente la complejidad.";
    }

    private String getComplexityByDepth(int depth) {
        switch (depth) {
            case 1: return "O(n)";
            case 2: return "O(n^2)";
            case 3: return "O(n^3)";
            case 4: return "O(n^4)";
            default: return "O(n^" + depth + ")";
        }
    }

    private String simplifyBigO(Map<String, Integer> complexities) {
        if (complexities.containsKey("O(n!)")) return "O(n!)";
        if (complexities.containsKey("O(2^n)")) return "O(2^n)";
        if (complexities.containsKey("O(n^4)")) return "O(n^4)";
        if (complexities.containsKey("O(n^3)")) return "O(n^3)";
        if (complexities.containsKey("O(n^2)")) return "O(n^2)";
        if (complexities.containsKey("O(n log n)")) return "O(n log n)";
        if (complexities.containsKey("O(n)")) return "O(n)";
        if (complexities.containsKey("O(log n)")) return "O(log n)";
        return "O(1)";
    }

    private String evaluateCodeQuality(String complexity) {
        switch (complexity) {
            case "O(1)": return "Excelente/Mejor";
            case "O(log n)": return "Bueno";
            case "O(n)": return "Aceptable";
            case "O(n log n)": return "Malo";
            case "O(n^2)": return "Pobre";
            case "O(n^3)": return "Muy pobre";
            case "O(n^4)": return "Extremadamente ineficiente";
            default: return "Horrible/Peor";
        }
    }

    // Método para obtener las complejidades detectadas
    public Map<String, Integer> getComplexities() {
        return complexitiesCount;
    }
}

File: pseudo/BigOAnalyzer.js
CLASE BigOAnalyzer

    // Inicializar mapa para contar complejidades
    INICIALIZAR complexitiesCount COMO un mapa vacío

    // Método para analizar código ingresado
    MÉTODO analyzeCode(código)
        SI el código no contiene "{" O "}"
            RETORNAR "El código no parece ser Java. Ingrese código válido."
        FIN SI

        DIVIDIR código en líneas
        INICIALIZAR loopDepth COMO 0
        LIMPIAR el mapa complexitiesCount

        // Análisis línea por línea
        PARA CADA línea EN el código
            REMOVER espacios en blanco de línea

            SI línea empieza con "for" O "while"
                INCREMENTAR loopDepth
                complexity ← getComplexityByDepth(loopDepth)
                AUMENTAR el conteo de complexity EN complexitiesCount
                AGREGAR resultado al análisis
            SINO SI línea ES "}"
                REDUCIR loopDepth A UN MÁXIMO DE 0
            FIN SI
        FIN PARA

        // Generar resumen de complejidades
        CREAR totalComplexities COMO una cadena vacía
        PARA CADA complejidad EN complexitiesCount
            AGREGAR formato "conteo x complejidad + " a totalComplexities
        FIN PARA
        REMOVER el último " + " de totalComplexities

        // Simplificar complejidad
        simplifiedComplexity ← simplifyBigO(complexitiesCount)

        // Evaluar calidad
        quality ← evaluateCodeQuality(simplifiedComplexity)

        RETORNAR análisis de línea + totalComplexities + simplifiedComplexity + quality
    FIN MÉTODO

    // Complejidad según profundidad
    MÉTODO getComplexityByDepth(profundidad)
        SEGÚN profundidad
            CASO 1: RETORNAR "O(n)"
            CASO 2: RETORNAR "O(n^2)"
            CASO 3: RETORNAR "O(n^3)"
            CASO 4: RETORNAR "O(n^4)"
            DEFAULT: RETORNAR "O(n^" + profundidad + ")"
        FIN SEGÚN
    FIN MÉTODO

    // Simplificar Big O
    MÉTODO simplifyBigO(complejidades)
        SI contiene "O(n!)": RETORNAR "O(n!)"
        SI contiene "O(2^n)": RETORNAR "O(2^n)"
        SI contiene "O(n^4)": RETORNAR "O(n^4)"
        SI contiene "O(n^3)": RETORNAR "O(n^3)"
        SI contiene "O(n^2)": RETORNAR "O(n^2)"
        SI contiene "O(n log n)": RETORNAR "O(n log n)"
        SI contiene "O(n)": RETORNAR "O(n)"
        SI contiene "O(log n)": RETORNAR "O(log n)"
        RETORNAR "O(1)"
    FIN MÉTODO

    // Calidad de complejidad
    MÉTODO evaluateCodeQuality(complejidad)
        SEGÚN complejidad
            CASO "O(1)": RETORNAR "Excelente/Mejor"
            CASO "O(log n)": RETORNAR "Bueno"
            CASO "O(n)": RETORNAR "Aceptable"
            CASO "O(n log n)": RETORNAR "Malo"
            CASO "O(n^2)": RETORNAR "Pobre"
            CASO "O(n^3)": RETORNAR "Muy pobre"
            CASO "O(n^4)": RETORNAR "Extremadamente ineficiente"
            DEFAULT: RETORNAR "Horrible/Peor"
        FIN SEGÚN
    FIN MÉTODO

    // Obtener las complejidades detectadas
    MÉTODO getComplexities()
        RETORNAR complexitiesCount
    FIN MÉTODO

FIN CLASE
